content/blog/insecure-software-development/index.md

```diff
@@ -0,0 +1,120 @@
+---
+title: Insecure Software Development
+date: "2022-01-28T08:00:00.284Z"
+description: "A tale of software development and hidden complexity"
+---
+
+## TL;DR;
+
+If you don't have the time or is not sure if you want to read this post, here are the main points discussed:
+
+1. Hidden complexity is a common cause of security vulnerabilities
+2. There are multiple ways to tackle complexity in codebases, each with different degrees of success and pitfalls
+3. Code review is a great tool to reduce complexity before it becomes a problem
+4. Type juggling is still a thing and is a good example of hidden complexity causing issues
+
+## Introduction
+
+Do you remember `Log4J`? What about `Dependency Confusion`? How could you not? At the time of this writing both happened over this past year and disrupted parts of our industry. Development and Infrastructure teams scrambled to patch their systems. Companies with poor asset management had it even worse, trying to identify and uncover vulnerable infrastructure while responding to these incidents. So what does both of these incidents
+have in common? I will blame _hidden complexity_ here.
+
+Log4J abstracted environment variables, JNDI, and other kinds of lookups in log messages through a custom syntax that is parsed at the time of logging. Dependency confusion was caused by package managers trying to abstract the domain of dependency management, which is far from simple. In both cases library maintainers made the behaviour implicit.
+
+## Complexity
+
+In practical terms we can define complexity as anything pertaining to our system that makes it difficult to understand or modify.
+
+As software evolves and gains more features, it becomes complex, with subtle coupling between components. This leads to bugs and longer development cycles. Eventually no single developer can keep the entire context of their codebases in their heads anymore and more bugs start popping up, slowing the development cycle even more. There are usually two approaches to fight complexity:
+
+1. Make the code simpler and obvious
+2. Encapsulate complexity, abstract
+
+### Make the code simpler and obvious
+
+The idea here is to make things _explicit_. The less our users have to guess in order to use what we built the better. Users here can range from developers to end users of our products, libraries or our own codebases.
+
+1. Good naming practices, meaningful names
+   - Can we understand on a high level what this code is trying to achieve without having to dig deeper?
+2. Eliminating special cases
+   - Can we enforce invariants?
+   - Can we just crash instead of handling special cases?
+   - Can we have a single implementation for our interfaces?
```

Comment:
> - Can we have a single implementation for our interfaces?

Is this referring to each interface having its own unique implementation rather than a more generalized and abstracted one? Or is this referring to a single implementation that is generalized and used through various interfaces?